cmake/extensions.cmake

```diff
@@ -0,0 +1,65 @@
+#
+# Copyright (c) 2020 Nordic Semiconductor ASA
+#
+# SPDX-License-Identifier: Apache-2.0
+#
+
+# cddl_generate(<target> <cddl_file> ENTRY_TYPES <entry_types> [VERBOSE])
+#
+# Add generated code to the project for decoding CBOR.
+#
+# <cddl_file> is a file with CDDL data describing the expected CBOR data.
+#
+# <entry_types> is a list of the types defined in the CDDL file for which there
+#               should be an exposed decoding function.
+#
+# Specify VERBOSE to add printing to the generated code. Check kconfig to add
+# printing in the low level CBOR decoding code.
+#
+# The result of the function is that a library is added to the project which
+# contains the generated decoding code. The code is generated at build time using
+# a script, and generates a c file and an h file with names derived from the
+# name of the <cddl_file>.
+function(cddl_generate target cddl_file)
```

Comment:
How about renaming this into: `target_cddl_sources`, that will make it follow CMake naming pattern, and code can the look as:
```
add_library(my_lib)

target_sources(my_lib PRIVATE my_sources.c)

target_cddl_sources(my_lib pet.cddl ENTRY_TYPES Pet)
target_cddl_sources(mylib serial_recovery.cddl ENTRY_TYPES Upload)
```

We could consider to also allow user to specify `<PUBLIC|PRIVATE|INTERFACE>` to really make it look similar to `target_sources`, but I will let you decide that. (Offering this could allow for additional use-cases, and I can describe those directly if you're curious)

Inspired from: https://github.com/oyvindronningstad/cddl_gen/blob/8adf96b300d4f80f85a8b6b7143f97ae867a4cd4/tests/cbor_decode/test3_simple/CMakeLists.txt#L13-L17

Reply:
This also match the behavior of a normal library, where `target_sources` will end up with following process during build time:
`target_sources(app file.c)` -->
```
file.c --> c-compiler  -> obj-files -> archiver -> libapp.c
```
`target_cddl_sources(app file.cddl)` -->
```
file.cddl --> cddl-compiler (generator)-> c-file -> c-compile -> obj-files -> archiver -> libapp.c
```

Note: asm stage is left out for simplicity.

Reply:
It doesn't support multiple sources per call though. Is `target_cddl_source` ok?

Reply:
> It doesn't support multiple sources per call though. Is target_cddl_source ok?

Good point, and if only allowing for one file, `target_cddl_source` is accepted.

But actually, if you make `PRIVATE` a multi value keyword, you could support multiple sources :smile: , so that one could do:
```
add_library(my_pets)
target_cddl_sources(my_pets PRIVATE pet.cddl dog.cddl cat.cddl ENTRY_TYPES Pet)
```
would that be a use-case ?

Reply:
I have to be a bit apologetic, but I am a bit cautious when we are adding new concepts for cmake library+sources very similar to the work done with **zephyr_library** and **zephyr_library_sources**

So... With **add_library** and **target_cddl_sources** we are showing that we are using regular CMAKE concepts, but with a twist... In my mind **target_sources_cddl** is a better naming scheme since it is clearer were we inherited from. You can even do a find with target_sources and it will still match

I think the clear benefit of postfixing cddl is that we signal that **target_sources_cddl** can be used from "any CMAKE library". It would also be possible to add a dummy function zephyr_library_sources_cddl if need be...

Reply:
It's tricky because ENTRY_TYPES must be set for each CDDL source. I can't think of a way to support that which isn't ugly :P.

It would have to be something like this

    target_sources_cddl(<target> [PRIVATE <source> ENTRY_TYPES <types>]

where everything inside brackets can be repeated.

Reply:
> It's tricky because ENTRY_TYPES must be set for each CDDL source. I can't think of a way to support that which isn't ugly :P.
> 

I see, I wasn't so much into the details.
I thought more file could share type, but as I learned now, they cannot, `target_cddl_source` is fine.

Btw, why is it `ENTRY_TYPES`, when you can only provide a single, shouldn't it be `ENTRY_TYPE` ?

Reply:
You can provide multiple types. It's done in one of the tests: https://github.com/oyvindronningstad/cddl_gen/pull/1/files#diff-e58f94dd4e1b3580e28261699c1251ebR12